### PR TITLE
[Spark load][Fe 1/5] Add spark etl job config

### PR DESCRIPTION
**Pull request by wyb**
Add spark etl job config, includes:
1. Schema of the load tables, including columns, partitions and rollups
2. Infos of the source file, including split rules, corresponding columns, and conversion rules
3. ETL output directory and file name format
4. Job properties
5. Version for further extension

https://github.com/apache/incubator-doris/issues/3010#issuecomment-635174675
#3433

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+ * {
+ * 	"tables": {
+ * 		10014: {
+ * 			"indexes": [{
+ * 			    "index_id": 10014,
+ * 				"columns": [{
+ * 				    "column_name": "k1",
+ * 				    "column_type": "SMALLINT",
+ * 				    "is_key": true,
+ * 				    "is_allow_null": true,
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java:41]: Add a field `defined_expr`, and default is empty. This is for materialized view 2.0, which column may has expr definition.

File: fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.io.Serializable;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+/** jobconfig.json file format
+ * {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java:32]: How about adding a `version` fields for further extension?

File: fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+ */
+public class EtlJobConfig implements Serializable {
+    // global dict
+    public static final String GLOBAL_DICT_TABLE_NAME = "doris_global_dict_table_%d";
+    public static final String DISTINCT_KEY_TABLE_NAME = "doris_distinct_key_table_%d_%s";
+    public static final String DORIS_INTERMEDIATE_HIVE_TABLE_NAME = "doris_intermediate_hive_table_%d_%s";
+
+    // hdfsEtlPath/jobs/dbId/loadLabel/PendingTaskSignature
+    private static final String ETL_OUTPUT_PATH_FORMAT = "%s/jobs/%d/%s/%d";
+    private static final String ETL_OUTPUT_FILE_NAME_DESC = "label.tableId.partitionId.indexId.bucket.schemaHash.parquet";
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java:132]: Add "version" field for further extension?

File: fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+ * 	"output_path": "hdfs://hdfs_host:port/user/output/10003/label1/1582599203397",
+ * 	"output_file_pattern": "label1.%d.%d.%d.%d.%d.parquet",
+ * 	"label": "label0",
+ * 	"properties": {
+ * 	    "strict_mode": false,
+ * 	    "timezone": "Asia/Shanghai"
+ * 	}
+ * }
+ */
+public class EtlJobConfig implements Serializable {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java:124]: Why this class should implement `Serializable`?
I think it is OK to convert to and from JSON.

File: fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+    public static final String ETL_OUTPUT_FILE_FORMAT = "parquet";
+
+    // dpp result
+    public static final String DPP_RESULT_NAME = "dpp_result.json";
+
+    public Map<Long, EtlTable> tables;
+    public String outputPath;
+    public String outputFilePattern;
+    public String label;
+    public EtlJobProperty properties;
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java:144]: If the variable name is changed, the serialized JSON will change.
Is there some method to seprate the variable name and serialized field name?

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java (hunk 4 above)
Comment (wyb) [reply fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java:124]: etl job failed if not implements Serializable interface. failed message: `org.apache.spark.SparkException: Task not serializable, object not serializable`

File: fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.annotations.SerializedName;
+
+import java.io.Serializable;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+/** jobconfig.json file format
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java:33]: You forgot to update the comment with the new format of json

File: fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+    }
+
+    public static class EtlColumnMapping implements Serializable {
+        @SerializedName(value = "functionName")
+        public String functionName;
+        @SerializedName(value = "args")
+        public List<String> args;
+        @SerializedName(value = "expr")
+        public String expr;
+        public Map<String, String> functionMap =
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java:502]: why not make it a `private static`?

File: fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+        }
+    }
+
+    public static class EtlColumnMapping implements Serializable {
+        @SerializedName(value = "functionName")
+        public String functionName;
+        @SerializedName(value = "args")
+        public List<String> args;
+        @SerializedName(value = "expr")
+        public String expr;
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/etl/EtlJobConfig.java:501]: better to add comment for this class, make it easy to understand the relation between functionName and expr

**Review by imay — APPROVED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**